### ISSUE TITLE
Extract reusable logic from RequireThisCheck

### ISSUE THREAD

**Issue by wilcoln** [checkstyle/checkstyle#8126]
If we take a look at source code of `RequireThisCheck`, we can see that it is a very long class with numerous methods and inner classes.

A lot of the code written in that class is not specific to the check and can be reused else where e.g. to solve https://github.com/checkstyle/checkstyle/issues/7895 and https://github.com/checkstyle/checkstyle/issues/7504

Essentially, the elements (methods and inner classes) that need to be extracted are of two types:
- Token related `->` should be moved to the already existing utility class `TokenUtil.java`
- Frame related `->` should be moved in a new utility class `FrameUtil.java`

**Comment by wilcoln** [comment]
This is issue has been created in response to https://github.com/checkstyle/checkstyle/pull/8049#pullrequestreview-391874442

**Comment by rnveach** [comment]
This is similar to duplicate of https://github.com/checkstyle/checkstyle/issues/5307 .

I started my own experience with this a while back. My branch is: https://github.com/rnveach/checkstyle/commits/frame_tracking_util

While I agree with this idea, I do question how much code from the check is re-usable right now and how much is specific to RequireThis. You can see me start to question this in some of my commits like https://github.com/rnveach/checkstyle/commit/e42500d5c20b9796d6720a56fcd10444b78e1274#diff-34b5b3acb6c71534624b1e3fa8e77376R335 where I started to mark methods as deprecated questioning if they were for a generic use or only for RequireThis and started adding new methods like https://github.com/rnveach/checkstyle/commit/e42500d5c20b9796d6720a56fcd10444b78e1274#diff-34b5b3acb6c71534624b1e3fa8e77376R218 . I thought I was renaming some too, but I couldn't find any.

I think we should make this a util rather than an abstract class. This will allow users to still extend their own custom classes without being forced into a strict class convention. This will allow the code to be portable from AbstractCheck to AbstractJavadocCheck too. As a utility, we may want to consider some type of caching to speed up multiple checks needing the frames for the same class.

I feel this class should have its own set of tests to prove it is working as expected and that it can't rely on RequireThis anymore if it is going to be its own standalone thing. I had my own examples of adding custom tests in https://github.com/rnveach/checkstyle/commit/d21f05f9466a37efdb20876622ef179c3d8e8872 .

I see 2 types of tests needed.

1) Displaying frames and what is being tracked for each frame to make sure we are capturing the look of the code (variables, fields, and methods). 
2) Checking if the "find" methods are returning the right locations and are not getting confused.

@romani Please review and let me know if you agree.

**Comment by romani** [comment]
> I think we should make this a util rather than an abstract class.

yes, or even just regular classes to be used by composition  rather than inheritance.

> I feel this class should have its own set of tests

it is very very last resort. It is better to create special pitest profile to test Checks together. But all tests should be on real Inputs (no fake tests). Let me know if you think that sticking to real testing is not feasible.

**Comment by rnveach** [comment]
I agree on real tests and was not talking about splitting pitest profile. We should give it real code and have it validate its internals. Other checks validate it's internals by checking for violations and no violations. This class will not create violations as it is more a utility. So it should be checking the results of its methods and it's captures isntead.

**Comment by romani** [comment]
> This class will not create violations as it is more a utility

the same for Utils classes, we try to test them from Checks. 

> So it should be checking the results of its methods and it's captures isntead.

I very strong in idea that we should avoid it as much as possible. We do not test methods. We test who such methods works on real code for real Checks. We deliver to user Checks, not Util methods, that is why we moved them(utils) out of API. I understand that life is life and there bunch of limitations and ..... but again pure UT on methods is absolute last resort, you need to prove that it is last resort in PR.

**Comment by rnveach** [comment]
I removed my comments on tests. Let me know if there is anything else you need to approve issue.

**Comment by romani** [comment]
@wilcoln , I approving issue, but please be caution with fix. Please read our comments on testing above, there will be hidden complications (to satisfy 100% coverage and pitest), please do not hesitate to pin us in PR to guide you on how to make testing (by pure UT or functional Test(full run)).

**Comment by wilcoln** [comment]
Understood, I will do my best.
